deps/spidershim/src/v8functiontemplate.cc

```diff
@@ -0,0 +1,315 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+
+#include <assert.h>
+
+#include "conversions.h"
+#include "v8local.h"
+#include "jsfriendapi.h"
+
+namespace {
+enum FunctionTemplatePrivateSlots {
+  CallbackSlot,            // Stores our FunctionCallback.
+  LengthSlot,              // Stores our length.
+  DataSlot,                // Stores our data value.
+  ProtoSlot,               // Stores our prototype's template, if any.
+  ParentSlot,              // Stores our FunctionTemplate parent, if any.
+  InstanceSlot,            // Stores our instantiation, if we have one.
+                           // XXXbz this doesn't seem quite right, but is
+                           // probably OK if we only support one global.
+  NumSlots
+};
+
+const js::Class functionTemplateClass = {
+  "FunctionTemplate", JSCLASS_HAS_RESERVED_SLOTS(NumSlots)
+};
+
+using namespace v8;
+
+void
+SetCallbackAndData(JSObject* templateObj, FunctionCallback callback,
+                   Handle<Value> data)
+{
+  assert(js::GetObjectClass(templateObj) == &functionTemplateClass);
+  if (callback) {
+    js::SetReservedSlot(templateObj, CallbackSlot,
+                        JS::PrivateValue(reinterpret_cast<void*>(callback)));
+  } else {
+    js::SetReservedSlot(templateObj, CallbackSlot, JS::UndefinedValue());
+  }
+
+  if (*data) {
+    js::SetReservedSlot(templateObj, DataSlot, *GetValue(data));
+  } else {
+    js::SetReservedSlot(templateObj, DataSlot, JS::UndefinedValue());
+  }
+}
+
+bool
+TemplateIsInstantiated(JSObject* templateObj)
+{
+  assert(js::GetReservedSlot(templateObj, InstanceSlot).isObject() ||
+         js::GetReservedSlot(templateObj, InstanceSlot).isUndefined());
+  return js::GetReservedSlot(templateObj, InstanceSlot).isObject();
+}
+} // anonymous namespace
+
+namespace v8 {
+
+// static
+Local<FunctionTemplate> FunctionTemplate::New(Isolate* isolate,
+                                              FunctionCallback callback,
+                                              Handle<Value> data,
+                                              Handle<Signature> signature,
+                                              int length)
+{
+  Local<Template> templ = Template::New(isolate, &functionTemplateClass);
+  if (*templ) {
+    return Local<FunctionTemplate>();
+  }
+
+  JSContext* cx = JSContextFromIsolate(isolate);
+  JS::RootedObject obj(cx, GetObject(*templ));
```

Comment:
FWIW this root here is not needed since anything wrapped in a `Local<>` is already rooted, and `SetCallbackAndData` doesn't require a `Handle`.